nexus/reconfigurator/preparation/src/lib.rs

```diff
@@ -165,6 +172,27 @@ impl PlanningInputFromDb<'_> {
         );
 
         let mut zpools_by_sled_id = {
+            // Gather all the datasets first, by Zpool ID
+            let mut datasets: Vec<_> = self
+                .dataset_rows
+                .iter()
+                .map(|dataset| {
+                    (
+                        ZpoolUuid::from_untyped_uuid(dataset.pool_id),
+                        dataset.clone(),
+                    )
+                })
+                .collect();
+            datasets.sort_by(|(a, _), (b, _)| a.cmp(&b));
```

Comment:
Nit - I think this could be

```suggestion
            datasets.sort_unstable_by_key(|(id, _)| id);
```

Reply:
Done in 60c3c1a520dbc7971bbe66073694f9fe9f18f8bb